feat: add support for `n install auto`, add `N_PREFIX` cache, replace node repo with `n`, fixes #6418

stasadev opened the pull request:
## The Issue

- #6418

## How This PR Solves The Issue

From https://www.npmjs.com/package/n#specifying-nodejs-versions:

There is an `auto` label to read the target version from a file in the current directory, or any parent directory. `n` looks for in order:

- `.n-node-version`: version on single line. Custom to `n`.
- `.node-version`: version on single line. Used by multiple tools: [node-version-usage](https://github.com/shadowspawn/node-version-usage)
- `.nvmrc`: version on single line. Used by `nvm`.
- if no version file found, look for `engine` as below.

The `engine` label looks for a `package.json` file and reads the `engines` field to determine compatible Node.js. Requires an installed version of `node`, and uses `npx semver` to resolve complex ranges.

---

This PR adds:
- cache for `n` just like it works for `nvm`.
- full compatibility for `n install`, as previously it was run from a `Dockerfile` with no mounted code, which limited its use.

## Manual Testing Instructions

Review the docs https://ddev--6420.org.readthedocs.build/en/6420/users/configuration/config/#nodejs_version

```
ddev config --nodejs-version=auto
echo "10" > .nvmrc
ddev restart
ddev logs -s web # to see how it was installed
ddev exec node --version
```

Repeat the same with different options from https://www.npmjs.com/package/n#specifying-nodejs-versions

```
ddev config --nodejs-version=engine
echo '{
    "engines": {
        "node": "14.20"
    }
}' > package.json
ddev restart
ddev logs -s web # to see how it was installed
ddev exec node --version
```

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

GuySartorelli commented on docs/content/users/configuration/config.md:
```diff
@@ -393,6 +393,9 @@ Node.js version for the web container’s “system” version.
 There is no need to configure `nodejs_version` unless you want a version other than the default version.
 
+!!!tip "Select the Node.js version from a file such as `.nvmrc` or `package.json`"
+    `nodejs_version` can be set to `auto`, which will use `n install auto` under the hood.
```
Probably worth either linking to `n` docs or explaining what that means. People who want this functionality may not know what `n` is or what `n install auto` does.

stasadev replied:
Thanks, that's a good suggestion.

GuySartorelli commented on pkg/ddevapp/templates.go:
```diff
@@ -70,7 +70,7 @@ const ConfigInstructions = `
 # Numeric version numbers can be complete (i.e. 18.15.0) or
 # incomplete (18, 17.2, 16). 'lts' and 'latest' can be used as well along with
-# other named releases.
+# other named releases, 'auto' reads the version from a file such as '.nvmrc' or 'package.json'.
```
In the pr description you mentioned an `engine` setting for checking `package.json`, but here you say `auto` will do that?

GuySartorelli replied:
May be worth mentioning the other files that it reads from (there's only 3 after all) to avoid people thinking it's only `.nvmrc`

stasadev replied:
I didn't even notice that I copied the part with `engine` 🙂, it doesn't work now, but it will.
I'm going to remove my hardcoded condition for `auto` so that it accepts anything that works for `n`.

GuySartorelli reviewed (commented):
Will test tomorrow, but glancing at it, this is a brilliant solution.

Couple things that stood out in the communication about the feature below.

stasadev commented on containers/ddev-webserver/ddev-webserver-prod-scripts/start.sh:
```diff
@@ -92,6 +92,10 @@ ls /var/www/html >/dev/null || (echo "/var/www/html does not seem to be healthy/
 mkdir -p /mnt/ddev-global-cache/{bashhistory/${HOSTNAME},mysqlhistory/${HOSTNAME},nvm_dir/${HOSTNAME},npm,yarn/classic,yarn/berry,corepack}
 
+if [ "${N_INSTALL_VERSION:-}" != "" ]; then
+  sudo n install "${N_INSTALL_VERSION}"
```
My mistake with `sudo`, thanks, and yes, it should also be available for prod/hardened, just like it works for `nvm`.

stasadev commented:
I have updated the code:

- Run `docker pull ddev/ddev-webserver:20240723_stasadev_n_install_auto` if you have already tested this PR.
- Fixed the issue where the node version was pulled every time by adding the `N_PREFIX` cache directory.
- Added `n install --offline`, so that it doesn't crash if you don't have internet.
- Updated the docs and added an example of what to do if you use `n install auto` and your file, for example `.nvmrc`, is not in the root of the project.

rfay commented on containers/ddev-webserver/ddev-webserver-base-files/etc/bashrc/commandline-addons.bashrc:
```diff
@@ -1,4 +1,6 @@
-export PATH="~/bin:$PATH:/var/www/html/vendor/bin:/var/www/html/bin:$DDEV_COMPOSER_ROOT/vendor/bin"
+PATH="$PATH:/var/www/html/vendor/bin:/var/www/html/bin:$DDEV_COMPOSER_ROOT/vendor/bin"
```
Perhaps slightly better

```suggestion
PATH="${PATH}:/var/www/html/vendor/bin:/var/www/html/bin:${DDEV_COMPOSER_ROOT}/vendor/bin"
```

However, I don't understand the removal of ~/bin. I think people might add things there with homeadditions.

stasadev replied:
> I don't understand the removal of `~/bin`

I didn't remove it, see the line below. I preserved the original order, so the binaries inside `N_PREFIX` will not override something in `~/bin`.

rfay commented on containers/ddev-webserver/ddev-webserver-base-scripts/start.sh:
```diff
@@ -93,8 +93,16 @@ ls /var/www/html >/dev/null || (echo "/var/www/html does not seem to be healthy/
+if command -v n >/dev/null 2>&1 && [ "${N_PREFIX:-}" != "" ] && [ "${N_INSTALL_VERSION:-}" != "" ]; then
+  ln -sf /mnt/ddev-global-cache/n_prefix/${HOSTNAME} "${N_PREFIX}"
+  # try a normal install that also uses cache and try again offline if it fails
+  n install "${N_INSTALL_VERSION}" || n install "${N_INSTALL_VERSION}" --offline
```
```suggestion
  n install "${N_INSTALL_VERSION}" || n install "${N_INSTALL_VERSION}" --offline || true
```

`n install --offline` fails if it doesn't find the version in the cache; I think we'd be better continuing.

stasadev replied:
I thought about this, but didn't implement it.

I think it's for good, I wish we could somehow notify the user of the error here.

Maybe I could check for `node --version` somewhere later in the `ddev start`, compare it with config and show a Go warning...

rfay commented on containers/ddev-webserver/ddev-webserver-prod-scripts/start.sh:
```diff
@@ -90,7 +90,14 @@ disable_xhprof
+if command -v n >/dev/null 2>&1 && [ "${N_PREFIX:-}" != "" ] && [ "${N_INSTALL_VERSION:-}" != "" ]; then
+  ln -sf /mnt/ddev-global-cache/n_prefix/${HOSTNAME} "${N_PREFIX}"
+  # try a normal install that also uses cache and try again offline if it fails
+  n install "${N_INSTALL_VERSION}" || n install "${N_INSTALL_VERSION}" --offline
```
```suggestion
  n install "${N_INSTALL_VERSION}" || n install "${N_INSTALL_VERSION}" --offline || true
```

rfay commented on docs/content/users/configuration/config.md:
```diff
@@ -385,13 +385,37 @@ Whether to skip mounting project into web container.
-Node.js version for the web container’s “system” version.
+Node.js version for the web container’s “system” version. [`n`](https://www.npmjs.com/package/n) tool is under the hood.
+
+There is no need to configure `nodejs_version` unless you want a version other than the default version.
```
```suggestion
There is no need to reconfigure `nodejs_version` unless you want a version other than the version already specified, which will be the default version at the time the project was configured.
```

rfay commented on docs/content/users/configuration/config.md:
```diff
@@ -385,13 +385,37 @@ Whether to skip mounting project into web container.
 
-There is no need to configure `nodejs_version` unless you want a version other than the default version.
+!!!tip "How to install the Node.js version from a file"
+    There is an `auto` label (see [full documentation](https://www.npmjs.com/package/n#specifying-nodejs-versions)):
```
```suggestion
Your project team may specify the Node.js version in a more general way than in the `.ddev/config.yaml`. For example, you may use a `.nvmrc` file, the `package.json`, or a similar technique. In that case, DDEV can use the external configuration provided by that file.

    There is an `auto` label (see [full documentation](https://www.npmjs.com/package/n#specifying-nodejs-versions)):
```

rfay commented on pkg/ddevapp/config.go:
```diff
@@ -973,8 +973,11 @@ func (app *DdevApp) RenderComposeYAML() (string, error) {
-		extraWebContent = extraWebContent + "\nRUN npm install -g n"
-		extraWebContent = extraWebContent + fmt.Sprintf("\nRUN n install %s && ln -sf /usr/local/bin/node /usr/local/bin/nodejs", app.NodeJSVersion)
+		extraWebContent = extraWebContent + fmt.Sprintf(`
+RUN npm install -g n
```
It's a little confusing that we try to do this at build time and then later during start.sh. But it seems reasonable.

```suggestion
RUN npm install -g n || true
```

stasadev replied:
This is a different thing, we run only `n install` during `start.sh`.

Maybe we should install it on `ddev-webserver` so it's available right away?

rfay replied:
I agree, we should have `n` already installed in ddev-webserver.

rfay commented on pkg/ddevapp/templates.go:
```diff
@@ -68,12 +68,24 @@ const ConfigInstructions = `
-# see https://www.npmjs.com/package/n#specifying-nodejs-versions
-# Note that you can continue using 'ddev nvm' or nvm inside the web container
-# to change the project's installed node version if you need to.
+#
+# See https://www.npmjs.com/package/n#specifying-nodejs-versions for the full documentation,
+# below are the most used options for the "nodejs_version":
+#
+# Numeric version numbers can be complete "4.9.1" or incomplete "8", with an optional leading v "v6.1".
+# "lts" is the newest Long Term Support official release; "latest" and "current" are the newest official release.
+#
+# There is an "auto" label to read the target version from a file in the DDEV_APPROOT directory,
+# or any parent directory. 'n' looks for in order:
+#   - ".n-node-version": version on single line. Custom to 'n'.
+#   - ".node-version": version on single line. Used by multiple tools https://github.com/shadowspawn/node-version-usage
+#   - ".nvmrc": version on single line. Used by 'nvm'.
+#   - if no version file found, look for "engine" as below.
+#
+# The "engine" label looks for a 'package.json' file and reads the engines field to determine compatible Node.js.
+#
+# Note that using of 'ddev nvm' is discouraged because "nodejs_version" is much easier to use,
+# can specify any version, and is more robust than using 'nvm'.
```
I think this section would be better as a link to the docs. It's probably too verbose here. We should probably reconsider most of what's here in template.go. I wonder if people find it useful.

stasadev replied:
Okay, a link to our docs sounds like a good idea to avoid this duplication.

rfay reviewed (changes requested):
Well, this is pretty awesome! I tested it and it worked without any questions at all. I have a number of minor suggestions. One of the things I really appreciate about you is how you study the possibilities and really understand them, rather than coming up with the most obvious "quick think" solution. All that conversation in the issue(s) and then you latched on to the excellent `n` techniques and adopted them. It's amazing work.

rfay commented on containers/ddev-webserver/Dockerfile:
```diff
@@ -55,6 +55,9 @@ RUN curl --fail -sSL https://raw.githubusercontent.com/netz98/n98-magerun/develo
 RUN curl --fail -sSL https://files.magerun.net/n98-magerun2-latest.phar -o /usr/local/bin/magerun2 && chmod 777 /usr/local/bin/magerun2
 RUN curl --fail -sSL https://raw.githubusercontent.com/netz98/n98-magerun2/develop/res/autocompletion/bash/n98-magerun2.phar.bash -o /etc/bash_completion.d/n98-magerun2.phar && chmod +x /usr/local/bin/magerun
 
+# n for nodejs_version
+RUN npm install -g n
+
```
I think it would be better to install `n` directly and then install nodejs using it. That way we don't have the /usr/bin/node and /usr/bin/npm hanging around and causing possible confusion.

rfay commented on containers/ddev-webserver/ddev-webserver-base-scripts/start.sh:
```diff
@@ -93,8 +93,28 @@ ls /var/www/html >/dev/null || (echo "/var/www/html does not seem to be healthy/
+sudo mkdir -p /mnt/ddev-global-cache/{bashhistory/${HOSTNAME},mysqlhistory/${HOSTNAME},n_prefix/${HOSTNAME},nvm_dir/${HOSTNAME},npm,yarn/classic,yarn/berry,corepack}
 sudo chown -R "$(id -u):$(id -g)" /mnt/ddev-global-cache/ /var/lib/php
+
+if [ "${N_PREFIX:-}" != "" ] && [ "${N_INSTALL_VERSION:-}" != "" ]; then
+  ln -sf /mnt/ddev-global-cache/n_prefix/${HOSTNAME} "${N_PREFIX}"
+  # try a normal install that also uses cache and try again offline if it fails
+  n_install_result=true
+  timeout 30 n install "${N_INSTALL_VERSION}" 2> >(tee /tmp/n-install-stderr.txt >&2) || n_install_result=false
+  if [ "${n_install_result}" = "false" ]; then
+    timeout 30 n install "${N_INSTALL_VERSION}" --offline 2> >(tee -a /tmp/n-install-stderr.txt >&2) && n_install_result=true
+  fi
+  if [ "${n_install_result}" = "true" ]; then
+    ln -sf ${N_PREFIX}/bin/node ${N_PREFIX}/bin/nodejs
+    # we don't need this file if everything is fine
+    rm -f /tmp/n-install-stderr.txt
+  else
+    # remove the symlink on error so that the system Node.js can be used
+    rm -f ${N_PREFIX}
+  fi
+  hash -r
+fi
```
Could we try moving this into a script that can be maintained separately? Like /usr/local/bin/n_install.sh or something?

rfay commented on pkg/ddevapp/ddevapp.go:
```diff
@@ -1473,6 +1473,16 @@ Fix with 'ddev config global --required-docker-compose-version="" --use-docker-c
 		util.Warning("Something is wrong with your Docker provider and /mnt/ddev_config is not mounted from the project .ddev folder. Your project cannot normally function successfully with this situation. Is your project in your home directory?")
 	}
 
+	if app.NodeJSVersion != nodeps.NodeJSDefault {
+		util.Debug(`checking nodejs_version: "%s" install for errors`, app.NodeJSVersion)
+		nInstallStderr, _, _ := app.Exec(&ExecOpts{
+			Cmd: "cat /tmp/n-install-stderr.txt || true",
+		})
+		if nInstallStderr != "" {
+			util.Warning("Unable to install nodejs_version: \"%s\", error output from `n install %s`:\n%s", app.NodeJSVersion, app.NodeJSVersion, nInstallStderr)
+		}
+	}
```
Consider generalizing this approach for other things that we skip over when offline, etc. Doesn't have to be done in this PR, but definitely worth thinking about. (This works quite nicely btw)

rfay reviewed (changes requested):
I tested what I think are the key changes from last time, and made a few suggestions. It works great.

stasadev commented:
Implemented the requested changes.

Do not forget to run `docker pull ddev/ddev-webserver:20240723_stasadev_n_install_auto` before testing.

stasadev commented:
Resolved conflicts, pushed `ddev-php-base` and `ddev-webserver`.

stasadev commented:
I didn't check at first, but this change for `n` adds +46Mb to the container.

@rfay, should I still add the change from the node repository to `n`?

ddev-php-base
![ddev-php-base](https://github.com/user-attachments/assets/fd825db7-6338-4c96-b1a2-4e38bd017e46)

ddev-webserver
![ddev-webserver](https://github.com/user-attachments/assets/75750447-1715-453f-ab08-459afa4013b1)

rfay commented:
Let's try to figure out where the size increase comes from. It doesn't make too much sense, since we're getting rid of node and npm, and moving them to a docker volume. There may be something that got lost. Thanks for paying attention. We do want to avoid increasing size this much on a PR like this.

stasadev commented:
Now everything is fine
![image](https://github.com/user-attachments/assets/b7a2d151-4d88-44ca-afd3-cef7feee852c)

stasadev commented on containers/ddev-webserver/Dockerfile:
```diff
@@ -165,7 +165,7 @@ RUN chmod -R 777 /var/log
 RUN mkdir -p /mnt/ddev-global-cache/mkcert /run/{php,blackfire} /var/cache/nginx /var/lib/nginx && chmod -R ugo+rw /mnt/ddev-global-cache/
 
-RUN chmod -fR ugo+w /usr/sbin /usr/bin /etc/nginx /var/cache/nginx /var/lib/nginx /run /var/www /etc/php/*/*/conf.d/ /var/lib/php/modules /etc/alternatives /usr/lib/node_modules /etc/php /etc/apache2 /var/log/apache2/ /var/run/apache2 /var/lib/apache2 /mnt/ddev-global-cache/*
+RUN chmod -fR ugo+w /usr/sbin /usr/bin /etc/nginx /var/cache/nginx /var/lib/nginx /run /var/www /etc/php/*/*/conf.d/ /var/lib/php/modules /etc/alternatives /usr/local/lib/node_modules /etc/php /etc/apache2 /var/log/apache2/ /var/run/apache2 /var/lib/apache2 /mnt/ddev-global-cache/*
```
Yes, `n` uses this folder.

GuySartorelli commented:
Sorry it took me so long to test this, but just tested now and it works exactly as expected.

I did notice that `ddev describe` will show `NodeJS:auto` instead of the version being used. It would be _nice_ if it showed the actual version in use... but on the other hand maybe saying "auto" here is better in some ways because it's telling the dev "it's not hardcoded in ddev configuration".
Either way having it say "auto" is the easier solution so probably no reason to change it.

rfay reviewed (approved):
This is great, I'm sure you didn't expect to spend this long on this. Thank you.

Testing, I note that if `auto` is set and nothing done to configure it, we see

```
Unable to install nodejs_version: "auto".
Error output from `n install auto`:

  Error: no file found for auto version (.n-node-version, .node-version, .nvmrc, or package.json)

     offline : matching cached versions

  Error: no file found for auto version (.n-node-version, .node-version, .nvmrc, or package.json)
```

node v20 is installed though. In a perfect world it would probably say "installing default nodejs_version 20" instead.  This is minor though and you're sick of this, so OK to get it in!

shadowspawn commented on containers/ddev-php-base/Dockerfile:
```diff
@@ -98,9 +96,14 @@ RUN apt-get -qq install --no-install-recommends --no-install-suggests -y \
-    nodejs \
     sqlite3
 
+RUN curl -fsSL https://raw.githubusercontent.com/tj/n/master/bin/n | bash -s "${NODE_VERSION}"
```
You could just save the curl download rather than piping it into bash, if that is a good enough install for your purposes. Basic idea:

```
curl -fsSL https://raw.githubusercontent.com/tj/n/master/bin/n -o /usr/local/bin/n
chmod a+x /usr/local/bin/n
n install "${NODE_VERSION}"
```

Longer explanation: https://github.com/tj/n/issues/809#issuecomment-2264129789

rfay replied:
If @shadowspawn says it, it must be true! (thanks for being here!) @stasadev had gone the direction of the documented `n` install-with-bash technique, and it required `make` to be used.

stasadev replied:
Thank you, @shadowspawn.

I will change this instruction to `n` when we prepare the next release of DDEV.

rfay commented:
Congrats all!
